alfresco-health-processor-platform/src/main/java/eu/xenit/alfresco/processor/service/CycleService.java

```diff
@@ -0,0 +1,134 @@
+            } catch (InterruptedException e) {
+                logger.error("Idling has failed, aborting...", e);
+                continueCycle.set(false);
+            }
+        }
+    }
+
+    void start(Cycle cycle) {
+        int txnLimit = cycle.getTxnLimit();
+        int timeIncrementSeconds = cycle.getTimeIncrementSeconds();
+        long firstCommitTime = cycle.getFirstCommitTime();
+
+        logger.debug("Tracking changes ... Start commit time: {}", LocalDateTime
+                .ofInstant(Instant.ofEpochMilli(firstCommitTime), ZoneId.systemDefault())
+                .toString());
+
+        processTxnRange(cycle, txnLimit, timeIncrementSeconds);
+
+        long timeIncrementEpoch = timeIncrementSeconds * 1000L;
+        while(txnHistoryIsCatchingUp(timeIncrementEpoch, cycle.getCommitTimeMs())) {
+            cycle.setCommitTimeMs(cycle.getCommitTimeMs() + timeIncrementEpoch);
+            processTxnRange(cycle, txnLimit, timeIncrementSeconds);
+        }
+    }
+
+    void processTxnRange(Cycle cycle, int txnLimit, int timeIncrementSeconds) {
+        // Save current progress in case of transactions collection failure
+        long maxTxId = cycle.getTransactionId();
+        long maxCommitTimeMs = cycle.getCommitTimeMs();
+
+        try {
+            List<Transaction> txs = getNodeTransactions(
+                    txnLimit, timeIncrementSeconds);
+
+            logger.debug("Found {} transactions", txs.size());
+
+            if(txs.size() > 0) {
+                // Yay, we have a list of transactions to process! Save them for later
+                maxCommitTimeMs = txs.stream().map(Transaction::getCommitTimeMs)
+                        .max(Long::compare).get();
+                maxTxId = txs.stream().map(Transaction::getId)
+                        .max(Long::compare).get();
+
+                for(Transaction tx : txs) {
+                    try {
+                        retryingTransactionHelper.doInTransaction(() -> {
+                            // do work
+                            return null;
+                        }, true, false);
+                    } catch (Exception e) {
+                        logger.error("Tracker loop failed: " + e.getMessage(), e);
+                    }
+                }
+            }
+
+            cycle.setTransactionId(maxTxId);
+            cycle.setCommitTimeMs( Long.max(maxCommitTimeMs, cycle.getCommitTimeMs()));
+
+        } catch (Exception ex) {
+            logger.error("Impossible to read tracker info: " + ex.getMessage(), ex);
+        }
+    }
+
+    private boolean txnHistoryIsCatchingUp(long timeIncrementEpoch, long commitTimeMs) {
+        long supposedLastScanTime = OffsetDateTime.now().toInstant().toEpochMilli() - timeIncrementEpoch;
+        return supposedLastScanTime > commitTimeMs;
+    }
+
+    private boolean reachedLastTx(long transactionId) {
+        return transactionId < 100_000;
```

Comment:
I'd suspect that this is checked against `cycle.getTxnLimit()`

Reply:
No, TxnLimit is how many txn's at a time to scan.
100k is replaced later by the max transaction from Solr in upcoming branch/change:
https://github.com/xenit-eu/alfresco-health-processor/blob/node-tx-service-solr/alfresco-health-processor-platform/src/main/java/eu/xenit/alfresco/processor/service/ProgressTracker.java#L29